geo/src/algorithm/affine_ops.rs

```diff
@@ -0,0 +1,242 @@
+    fn default() -> Self {
+        Self::identity()
+    }
+}
+
+impl<T: CoordNum> AffineTransform<T> {
+    /// Create a new affine transformation by composing two `AffineTransform`s
+    pub fn compose(&self, other: &AffineTransform<T>) -> Self {
+        // lol
+        Self([
+            [
+                (self.0[0][0] * other.0[0][0])
+                    + (self.0[0][1] * other.0[1][0])
+                    + (self.0[0][2] * other.0[2][0]),
+                (self.0[0][0] * other.0[0][1])
+                    + (self.0[0][1] * other.0[1][1])
+                    + (self.0[0][2] * other.0[2][1]),
+                (self.0[0][0] * other.0[0][2])
+                    + (self.0[0][1] * other.0[1][2])
+                    + (self.0[0][2] * other.0[2][2]),
+            ],
+            [
+                (self.0[1][0] * other.0[0][0])
+                    + (self.0[1][1] * other.0[1][0])
+                    + (self.0[1][2] * other.0[2][0]),
+                (self.0[1][0] * other.0[0][1])
+                    + (self.0[1][1] * other.0[1][1])
+                    + (self.0[1][2] * other.0[2][1]),
+                (self.0[1][0] * other.0[0][2])
+                    + (self.0[1][1] * other.0[1][2])
+                    + (self.0[1][2] * other.0[2][2]),
+            ],
+            [
+                (self.0[2][0] * other.0[0][0])
+                    + (self.0[2][1] * other.0[1][0])
+                    + (self.0[2][2] * other.0[2][0]),
+                (self.0[2][0] * other.0[0][1])
+                    + (self.0[2][1] * other.0[1][1])
+                    + (self.0[2][2] * other.0[2][1]),
+                (self.0[2][0] * other.0[0][2])
+                    + (self.0[2][1] * other.0[1][2])
+                    + (self.0[2][2] * other.0[2][2]),
+            ],
+        ])
+    }
+    /// Create the identity matrix
+    pub fn identity() -> Self {
+        Self([
+            [T::one(), T::zero(), T::zero()],
+            [T::zero(), T::one(), T::zero()],
+            [T::zero(), T::zero(), T::one()],
+        ])
+    }
+
+    /// Create an affine transform for scaling, scaled by factors along the `x` and `y` dimensions.
+    /// The point of origin is *usually* given as the 2D bounding box centre of the geometry, but
+    /// any coordinate may be specified.
+    /// Negative scale factors will mirror or reflect coordinates
+    pub fn scaled(mut self, xfact: T, yfact: T, origin: Point<T>) -> Self {
+        let (x0, y0) = origin.x_y();
+        let xoff = x0 - (x0 * xfact);
+        let yoff = y0 - (y0 * yfact);
+        self.0[0][0] = self.0[0][0] * xfact;
+        self.0[1][1] = self.0[1][1] * yfact;
+        self.0[0][2] = xoff;
+        self.0[1][2] = yoff;
+        self
+    }
+
+    /// Create an affine transform for translation, shifted by offsets along the `x` and `y` dimensions
+    pub fn translated(mut self, xoff: T, yoff: T) -> Self {
+        // first row
+        self.0[0][0] = T::one();
+        self.0[0][1] = T::zero();
```

Comment:
Given the "builder-ish" API from signature, I had assumed I'd be able to chain methods together for composition. 

e.g. this will currently compile (I think):
```
let transform = AffineTransform::default()
    .scaled(10.0, 10.0, Point(Coordinate::zero())
    .translated(2.0, 4.0)
    .rotated(45.0);

polygon.map_coords(|c| transform.apply)
```

I assumed this would give me an output which had been scaled, translated, and rotated.

But I think as phrased it doesn't behave that way - the later calls replace (rather than compose wth) some of the previous values.

I'd say ultimately it'd be great to support this builder pattern style API for AffineTransforms, but alternatively, if you don't want to mess around with that for now, it might be slightly simpler to offer constructors rather than mutable methods. Then at least they wouldn't *look* like they were composing.

e.g. something like this might avoid the above confusion:
```diff
// constructor, not method
/// AffineTransform::translate(x, y).apply(coord)
- pub fn translated(mut self, xoff: T, yoff: T) -> Self {
+ pub fn translate(xoff: T, yoff: T) -> Self {
+     let mut transform = Self::identity();
      // first row
-     self.0[0][0] = T::one();
-     self.0[0][1] = T::zero();
-     self.0[0][2] = xoff;
+     transform.0[0][2] = xoff;
      // second row
-     self.0[1][0] = T::zero();
-     self.0[1][1] = T::one();
-     self.0[1][2] = yoff;
+     transform.0[1][2] = yoff;

-   self
+   transform
}
```

Once methods like that exist, I think creating an intuitive composable builder pattern (possibly in a follow up PR) would be pretty straight forward:
```rust
pub fn translated(mut self, xoff: T, yoff: T) -> Self {
    self.compose(AffineTransform::translate(xoff, yoff))
}
```

I have a similar concern for the other transforms - `scaled`, `rotated`, `skewed`.

Reply:
As an extreme example - I think even this should be possible:

```rust
let transform = AffineTransform::identity()
    .scaled(2.0, 2.0, Point(Coordiante::zero()))
    .scaled(2.0, 2.0, Point(Coordiante::zero())).
```

And would expect it to achieve a 4x scaling.  I admit I wouldn't expect someone to write the above code *exactly* but something more reasonable which the currently proposed API doesn't easily support:

```rust
struct MapCanvasState {
    current_transform: AffineTransform
}
impl MapCanvasState {
    fn zoom_in(&mut self, origin: Point<T>) {
        self.current_transform = self.current_transform.scale(2.0, 2.0, origin)
    }
}
```

Reply:
Ohhhh no, your examples are _exactly_ what I would want (I just haven't really messed about with it yet bc I've been too busy assembling it), and that's why I wanted `compose` – let me try it by using `compose` inside the calls.

I wonder whether a `reset` method (returns matrix to the identity) might be useful too…

Reply:
I've now re-written all the mutate-in-place ops to the following format (note assignment of new matrix to `self.0`)

```Rust
pub fn scaled(mut self, xfact: T, yfact: T, origin: Point<T>) -> Self {
    let (x0, y0) = origin.x_y();
    let xoff = x0 - (x0 * xfact);
    let yoff = y0 - (y0 * yfact);
    self.0 = self
        .compose(&AffineTransform::new(
            xfact,
            T::zero(),
            T::zero(),
            yfact,
            xoff,
            yoff,
        ))
        .0;
    self
}
```

and written a test for cumulative ops, which are now behaving as expected. E.g.:

```Rust
let p0 = Point::new(0.0f64, 0.0);
// scale once
let mut scale_a = AffineTransform::default().scaled(2.0, 2.0, p0);
// rotate
scale_a = scale_a.rotated(45.0, p0);
// rotate back
scale_a = scale_a.rotated(-45.0, p0);
// scale up again, doubling
scale_a = scale_a.scaled(2.0, 2.0, p0);
```

Two issues:

1. Do we need the constructors?
2. Do we need to call out the fact that chaining ops on geometries isn't efficient? E.g.

`geom.scale(…).rotate(…)` will produce two `MapCoords` ops. I don't think there's a good way to avoid _that_, but we could a) document the perf pitfall, and b) provide a `Transform` trait for `MapCoords<T, T, Output = G> + MapCoordsInPlace<T>` so a cumulative `AffineTransform` can be built and passed to it. WDYT?

Reply:
> I've now re-written all the mutate-in-place ops

Nice!

> Do we need the constructors?

I think `AffineTransform::scale(2.0, 2.0, point)` | `AffineTransform::scale(2.0, 2.0, point).translated(1.0, 1.0)` would be slightly nicer than `AffineTransform::default().scaled(2.0, 2.0, point)` but I don't feel that it's *necessary* and could easily be added later.

>  provide a Transform trait [..] so a cumulative AffineTransform can be built and passed to it. 

I'd like to see that! I'd be OK with it getting added to this PR or in a followup PR. We already have `proj::Transform` in our geo namespace, so it should probably be called something else. I was calling it `trait AffineTransformable` in my [mockup](https://github.com/georust/geo/issues/728#issuecomment-1061078327), but I'm not married to that.

Reply:
Great. I'll open an issue for follow-up work, and broadcast this in case anyone else wants to kick the tires.